apps/ejabberd/src/mod_push.erl

```diff
@@ -0,0 +1,291 @@
+%%%-------------------------------------------------------------------
+%%% @author Rafal Slota
+%%% @copyright (C) 2017 Erlang Solutions Ltd.
+%%% This software is released under the Apache License, Version 2.0
+%%% cited in 'LICENSE.txt'.
+%%% @end
+%%%-------------------------------------------------------------------
+%%% @doc
+%%% Implementation of XEP-0357
+%%% @end
+%%%-------------------------------------------------------------------
+-module(mod_push).
+-author('rafal.slota@erlang-solutions.com').
+-behavior(gen_mod).
+-xep([{xep, 357}, {version, "0.2.1"}]).
+
+-include_lib("ejabberd/include/ejabberd.hrl").
+-include_lib("ejabberd/include/jlib.hrl").
+
+%%--------------------------------------------------------------------
+%% Exports
+%%--------------------------------------------------------------------
+
+%% gen_mod handlers
+-export([start/2, stop/1]).
+
+%% Hooks and IQ handlers
+-export([iq_handler/3,
+         handle_publish_response/4,
+         filter_packet/1,
+         remove_user/2]).
+
+%% Types
+-export_type([pubsub_node/0, form_field/0, form/0]).
+
+%%--------------------------------------------------------------------
+%% Definitions
+%%--------------------------------------------------------------------
+
+-define(PUSH_FORM_TYPE, <<"urn:xmpp:push:summary">>).
+
+-callback init(Host :: ejabberd:server(), Opts :: list()) -> ok.
+-callback enable(UserJID :: ejabberd:jid(), PubsubJID :: ejabberd:jid(),
+                 Node :: mod_push:pubsub_node(), Form :: mod_push:form()) ->
+    ok | {error, Reason :: term()}.
+-callback disable(UserJID :: ejabberd:jid(), PubsubJID :: ejabberd:jid(),
+                  Node :: mod_push:pubsub_node()) -> ok | {error, Reason :: term()}.
+-callback get_publish_services(User :: ejabberd:jid()) ->
+    {ok, [{PubSub :: ejabberd:jid(), Node :: mod_push:node(), Form :: mod_push:form()}]} |
+    {error, Reason :: term()}.
+
+%% Types
+-type pubsub_node()        :: binary().
+-type form_field()  :: {Name :: binary(), Value :: binary()}.
+-type form()        :: [form_field()].
+
+%%--------------------------------------------------------------------
+%% Module callbacks
+%%--------------------------------------------------------------------
+
+-spec start(Host :: ejabberd:server(), Opts :: list()) -> any().
+start(Host, Opts) ->
+    ?INFO_MSG("mod_push starting on host ~p", [Host]),
+
+    ok = application:ensure_started(worker_pool),
```

Comment:
I'd avoid command that could start `worker_pool` here, as it should be done during VM boot. I think it would be better to check for `wpool:start_sup_pool` return value to not only verify that `worker_pool` is running but also that pool creation was indeed successful.